src/Type/Php/FilterInputDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,63 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\ConstFetch;
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\MixedType;
+use PHPStan\Type\NeverType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+use function count;
+use function in_array;
+
+class FilterInputDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function __construct(private FilterFunctionReturnTypeHelper $filterFunctionReturnTypeHelper)
+	{
+	}
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'filter_input';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): ?Type
+	{
+		if (count($functionCall->getArgs()) < 2) {
+			return null;
+		}
+
+		$typeArgExpr = $functionCall->getArgs()[0]->value;
+		$varNameType = $scope->getType($functionCall->getArgs()[1]->value);
+		$varNameTypeIsString = $varNameType->isString();
+		if (
+			$varNameTypeIsString->no()
+			|| !($typeArgExpr instanceof ConstFetch)
```

Comment:
Still doesn't feel like a good way to handle those constants though. Can this be done better? E.g. what if people pass constant ints or a general integer.

Reply:
decided to simplify some sanity checks. the preciser type is only determined when a constant is passed, but if not it doesn't assume a fatal error anymore, but simply falls back to the default return type. there are too many weird PHP-version-dependent edge cases most likely nobody is interested it. e.g. https://3v4l.org/OH6Nt#veol

it can be still made stricter / preciser any time in the future